In [24]:
import sympy as sp

### Задание 1.4. Методы минимальных путей и сечений ###

#### *1. Вычислите структурную функцию работоспособности (СФР) методами минимальных путей и сечений для участка газотранспортной сети Ямбург-Челябинск.* ####

<p align="center">
  <img src="images\9.png" width=700/>
</p>

Построим схему сети:

<p align="center">
  <img src="images\10.jpg" width=700/>
</p>

**Обозначения:**
1. Ямбург - развилка вост.
2. Ямбург - Уренгой
3. Уренгой - развилка вост.
4. Уренгой - Челябинск
5. рзавилка вост. - Надым - Пунга - развилка зап.
6. развилка вост. - развилка зап.
7. развилка зап. - Челябинск

**Минимальные пути:**

2 - 3 - 5 - 7

2 - 3 - 6 - 7

1 - 3 - 4

1 - 5 - 7

1 - 6 - 7

2 - 4

$$ S(x) = 1 - \prod_{i=1}^{6} \alpha_i(x) = 1 - (1-x_2 x_3 x_5 x_7)(1 - x_2 x_3 x_6 x_7)(1 - x_1 x_3 x_4)(1 - x_1 x_5 x_7)(1 - x_1 x_6 x_7)(1 - x_2 x_4) $$

**Минимальные сечения:**

1 - 2

5 - 6 - 4

7 - 4

1 - 3 - 4

2 - 3 - 5 - 6

2 - 3 - 7

$$ S(x) = \prod_{k=1}^{6} \beta_k(x) = (1 - (1 - x_1)(1 - x_2))(1 - (1 - x_4)(1 - x_5)(1 - x_6))(1 - (1 - x_4)(1 - x_7))\cdot\\\cdot(1 - (1 - x_1)(1 - x_3)(1 - x_4))(1 - (1 - x_2)(1 - x_3)(1 - x_5)(1 - x_6))(1 - (1 - x_2)(1 - x_3)(1 - x_7))





#### *2. Методом минимальных путей и сечений вычислите верхнюю и нижнюю оценки вероятности безотказной работы сети Ямбург-Челябинск в течение 3-х лет, если известна протяженность участков (табл.), а удельная интенсивность отказов одной нитки газопровода – 0.4 отказа на тыс. км в год.* ####

<p align="center">
  <img src="images\11.jpg" width=700/>
</p>

$$ \prod_{i=1}^{n} \mathbb{P}\{\beta_i(x) = 1\} \leqslant R \leqslant 1 - \prod_{i=1}^{n}\mathbb{P}\{\alpha_i(x) = 0\} $$

**Представим схему в виде:**

Параллельном

<p align="center">
  <img src="images\12.jpg" width=700/>
</p>


Последовательном

<p align="center">
  <img src="images\13.jpg" width=700/>
</p>



In [25]:
p = sp.symbols('p_1:8')
S_min = sp.Function('S_min')
S_max = sp.Function('S_max')
S_12 = 1-(1-p[0])*(1-p[1])
S_456 = 1-(1-p[3])*(1-p[4])*(1-p[5])
S_47 = 1-(1-p[3])*(1-p[6])
S_134 = 1-(1-p[0])*(1-p[2])*(1-p[3])
S_2356 = 1-(1-p[1])*(1-p[2])*(1-p[4])*(1-p[5])
S_237 = 1-(1-p[1])*(1-p[2])*(1-p[6])

In [26]:
S_min = S_12 * S_456 * S_47 * S_134 * S_2356 * S_237
S_min_val = S_min.subs({p[0]: 1-(1-sp.exp(-0.4*10**(-3)*160*3))**3, 
            p[1]: 1-(1-sp.exp(-0.4*10**(-3)*140*3)),
            p[2]: 1-(1-sp.exp(-0.4*10**(-3)*130*3))**3,
            p[3]: 1-(1-sp.exp(-0.4*10**(-3)*(1200+580)*3))**2, 
            p[4]: 1-(1-sp.exp(-0.4*10**(-3)*(240+500+280)*3))**3, 
            p[5]: 1-(1-sp.exp(-0.4*10**(-3)*1000*3))**4,
            p[6]: (1-(1-sp.exp(-0.4*10**(-3)*130*3))**4)*(1-(1-sp.exp(-0.4*10**(-3)*750*3))**2)})

print("S_min = ", S_min_val)

S_min =  0.677837232603970


In [27]:
S_2357 = 1-p[1]*p[2]*p[4]*p[6]
S_2367 = 1-p[1]*p[2]*p[5]*p[6]
S_24 = 1-p[1]*p[3]
S_134 = 1-p[0]*p[2]*p[3]
S_157 = 1-p[0]*p[4]*p[6]
S_167 = 1-p[0]*p[5]*p[6]

In [28]:
S_max = 1 - S_2357 * S_2367 * S_24 * S_134 * S_157 * S_167
S_max_val = S_max.subs({p[0]: 1-(1-sp.exp(-0.4*10**(-3)*160*3))**3, 
            p[1]: 1-(1-sp.exp(-0.4*10**(-3)*140*3)),
            p[2]: 1-(1-sp.exp(-0.4*10**(-3)*130*3))**3,
            p[3]: 1-(1-sp.exp(-0.4*10**(-3)*(1200+580)*3))**2, 
            p[4]: 1-(1-sp.exp(-0.4*10**(-3)*(240+500+280)*3))**3, 
            p[5]: 1-(1-sp.exp(-0.4*10**(-3)*1000*3))**4,
            p[6]: (1-(1-sp.exp(-0.4*10**(-3)*130*3))**4)*(1-(1-sp.exp(-0.4*10**(-3)*750*3))**2)})

print("S_max = ", S_max_val)

S_max =  0.929047109323050


$$ 0.528416304079905 \leqslant R \leqslant 0.929047109323050

#### *3. Решите предыдущую задачу методом рёберно-непересекающихся минимальных путей и сечений.* ####

<style>
table {
  border: none;
}
td {
  border: none;
}
</style>

<table style="border: none;">
  <tr>
    <td style="vertical-align: top; text-align: center; border: none;">Реберно-непересекающиеся пути</td>
    <td style="vertical-align: top; text-align: center; border: none;">Реберно-непересекающиеся сечения</td>
  </tr>
  <tr>
    <td style="vertical-align: top; text-align: center; border: none;">
      <img src="images/14.jpg" width="300" />
    </td>
    <td style="vertical-align: top; text-align: center; border: none;">
      <img src="images/15.jpg" width="400" />
    </td>
  </tr>
</table>

$$ max(S_{12457}, S_{12467}, S_{2367}, S_{2357}, S_{134}) < R < min(S_{1247}, S_{12456}, S_{134}, S_{2356}, S_{237})


In [29]:
p = sp.symbols('p_1:8')
S_12457 = 1 - (1-p[0]*p[4]*p[6])*(1-p[1]*p[3])
S_12467 = 1 - (1-p[0]*p[5]*p[6])*(1-p[1]*p[3])
S_2367 = p[1]*p[2]*p[5]*p[6]
S_2357 = p[1]*p[2]*p[4]*p[6]
S_134 = p[0]*p[2]*p[3]

In [30]:
values = {p[0]: 1-(1-sp.exp(-0.4*10**(-3)*160*3))**3, 
            p[1]: 1-(1-sp.exp(-0.4*10**(-3)*140*3)),
            p[2]: 1-(1-sp.exp(-0.4*10**(-3)*130*3))**3,
            p[3]: 1-(1-sp.exp(-0.4*10**(-3)*(1200+580)*3))**2, 
            p[4]: 1-(1-sp.exp(-0.4*10**(-3)*(240+500+280)*3))**3, 
            p[5]: 1-(1-sp.exp(-0.4*10**(-3)*1000*3))**4,
            p[6]: (1-(1-sp.exp(-0.4*10**(-3)*130*3))**4)*(1-(1-sp.exp(-0.4*10**(-3)*750*3))**2)}

In [31]:
S_min = max(S_12457.subs(values), S_12467.subs(values), S_2367.subs(values), S_2357.subs(values), S_134.subs(values))
S_min

0.586253132734936

In [32]:
S_1247 = (1-(1-p[0])*(1-p[1]))*(1-(1-p[3])*(1-p[6]))
S_12456 = (1-(1-p[0])*(1-p[1]))*(1-(1-p[3])*(1-p[4])*(1-p[5]))
S_134 = 1-(1-p[0])*(1-p[2])*(1-p[3])
S_2356 = 1-(1-p[1])*(1-p[2])*(1-p[4])*(1-p[5])
S_237 = 1-(1-p[1])*(1-p[2])*(1-p[6])

In [33]:
S_max = min(S_1247.subs(values), S_12456.subs(values), S_134.subs(values), S_2356.subs(values), S_237.subs(values))
S_max

0.725307360782395

$$ 0.586253132734936 < R < 0.725307360782395

#### *4. Преобразовав СФР к виду, линейному по каждому аргументу, вычислите точную вероятность безотказной работы участка сети, рассмотренного в предыдущих задачах.* ####

$$ R = S(p) $$

In [34]:
x = sp.symbols('x_1:8')
S_2357 = 1 - x[1]* x[2]* x[4]*x[6]
S_2367 = 1 - x[1]* x[2]* x[5]*x[6]
S_24  = 1 - x[1]* x[3]
S_134 = 1 - x[0]* x[2]* x[3]
S_157 = 1 - x[0]* x[4]* x[6]
S_167 = 1 - x[0]* x[5]* x[6]

In [35]:
x = sp.symbols('x_1:8')
S_2357 = 1 - x[1]* x[2]* x[4]*x[6]
S_2367 = 1 - x[1]* x[2]* x[5]*x[6]
S_24  = 1 - x[1]* x[3]
S_134 = 1 - x[0]* x[2]* x[3]
S_157 = 1 - x[0]* x[4]* x[6]
S_167 = 1 - x[0]* x[5]* x[6]

In [36]:
S = sp.Function('S')(x[0], x[1], x[2], x[3], x[4], x[5], x[6])
S = sp.Eq(S, 1 - S_2357 * S_2367 * S_24 * S_134 * S_157 * S_167)
S

Eq(S(x_1, x_2, x_3, x_4, x_5, x_6, x_7), -(-x_2*x_4 + 1)*(-x_1*x_3*x_4 + 1)*(-x_1*x_5*x_7 + 1)*(-x_1*x_6*x_7 + 1)*(-x_2*x_3*x_5*x_7 + 1)*(-x_2*x_3*x_6*x_7 + 1) + 1)

In [37]:
temps = (1 - S_2357 * S_2367 * S_24 * S_134 * S_157 * S_167).expand()
temps

-x_1**3*x_2**3*x_3**3*x_4**2*x_5**2*x_6**2*x_7**4 + x_1**3*x_2**2*x_3**3*x_4*x_5**2*x_6**2*x_7**4 + x_1**3*x_2**2*x_3**2*x_4**2*x_5**2*x_6*x_7**3 + x_1**3*x_2**2*x_3**2*x_4**2*x_5*x_6**2*x_7**3 - x_1**3*x_2*x_3**2*x_4*x_5**2*x_6*x_7**3 - x_1**3*x_2*x_3**2*x_4*x_5*x_6**2*x_7**3 - x_1**3*x_2*x_3*x_4**2*x_5*x_6*x_7**2 + x_1**3*x_3*x_4*x_5*x_6*x_7**2 + x_1**2*x_2**3*x_3**3*x_4**2*x_5**2*x_6*x_7**3 + x_1**2*x_2**3*x_3**3*x_4**2*x_5*x_6**2*x_7**3 + x_1**2*x_2**3*x_3**2*x_4*x_5**2*x_6**2*x_7**4 - x_1**2*x_2**2*x_3**3*x_4*x_5**2*x_6*x_7**3 - x_1**2*x_2**2*x_3**3*x_4*x_5*x_6**2*x_7**3 - x_1**2*x_2**2*x_3**2*x_4**2*x_5**2*x_7**2 - 2*x_1**2*x_2**2*x_3**2*x_4**2*x_5*x_6*x_7**2 - x_1**2*x_2**2*x_3**2*x_4**2*x_6**2*x_7**2 - x_1**2*x_2**2*x_3**2*x_5**2*x_6**2*x_7**4 - x_1**2*x_2**2*x_3*x_4*x_5**2*x_6*x_7**3 - x_1**2*x_2**2*x_3*x_4*x_5*x_6**2*x_7**3 + x_1**2*x_2*x_3**2*x_4*x_5**2*x_7**2 + 2*x_1**2*x_2*x_3**2*x_4*x_5*x_6*x_7**2 + x_1**2*x_2*x_3**2*x_4*x_6**2*x_7**2 + x_1**2*x_2*x_3*x_4**2*x_5*x_7 + x_1

In [38]:
for i in range(7):
    for j in range(4, 1, -1):
        temp = temps.replace(x[i]**j, x[i])
        temps = temp
temps

-2*x_1*x_2*x_3*x_4*x_5*x_6*x_7 + 2*x_1*x_2*x_3*x_4*x_5*x_7 + 2*x_1*x_2*x_3*x_4*x_6*x_7 - x_1*x_2*x_3*x_4 + x_1*x_2*x_3*x_5*x_6*x_7 - x_1*x_2*x_3*x_5*x_7 - x_1*x_2*x_3*x_6*x_7 + x_1*x_2*x_4*x_5*x_6*x_7 - x_1*x_2*x_4*x_5*x_7 - x_1*x_2*x_4*x_6*x_7 + x_1*x_3*x_4*x_5*x_6*x_7 - x_1*x_3*x_4*x_5*x_7 - x_1*x_3*x_4*x_6*x_7 + x_1*x_3*x_4 - x_1*x_5*x_6*x_7 + x_1*x_5*x_7 + x_1*x_6*x_7 + x_2*x_3*x_4*x_5*x_6*x_7 - x_2*x_3*x_4*x_5*x_7 - x_2*x_3*x_4*x_6*x_7 - x_2*x_3*x_5*x_6*x_7 + x_2*x_3*x_5*x_7 + x_2*x_3*x_6*x_7 + x_2*x_4

In [39]:
temps.subs({x[0]: 1-(1-sp.exp(-0.4*10**(-3)*160*3))**3, 
            x[1]: 1-(1-sp.exp(-0.4*10**(-3)*140*3)),
            x[2]: 1-(1-sp.exp(-0.4*10**(-3)*130*3))**3,
            x[3]: 1-(1-sp.exp(-0.4*10**(-3)*(1200+580)*3))**2, 
            x[4]: 1-(1-sp.exp(-0.4*10**(-3)*(240+500+280)*3))**3, 
            x[5]: 1-(1-sp.exp(-0.4*10**(-3)*1000*3))**4,
            x[6]: (1-(1-sp.exp(-0.4*10**(-3)*130*3))**4)*(1-(1-sp.exp(-0.4*10**(-3)*750*3))**2)})

0.683043062558387

#### *5. Напишите программу, которая вычисляет СФР методами минимальных путей и минимальных сечений для произвольного графа.* ####

In [40]:
import sympy as sym
import itertools
from itertools import combinations

А) Сечения

Удаляет дубликаты подсписков из списка подсписков

In [41]:
def remove_duplicate_sublists(lst):
    seen = set()
    result = []

    for sublist in lst:
        sublist_tuple = tuple(sorted(sublist))
        if sublist_tuple not in seen:
            seen.add(sublist_tuple)
            result.append(list(sublist_tuple))

    return result

Удаляет дубликаты подсписков в списке lst. Преобразует подсписки в кортежи для удобного сравнения и сохраняет только уникальные

Проверка корректности разбиения вершин графа на два множества

In [42]:
def is_correct(set1,set2,graph,start,end):
    for i in set1:
        k=0
        for j in set1:
            if(str(i) in graph[str(j)] or str(j) in graph[str(i)]):
                k+=1
        if k==0 and len(set1)>1:
            return False
    for i in set2:
        k=0
        for j in set2:
            if(str(i) in graph[str(j)] or str(j) in graph[str(i)]):
                k+=1
        if k==0 and len(set2)>1:
            return False
    if(int(start) in set1 and int(end) not in set2):
        return False
    if(int(start) in set2 and int(end) not in set1):
        return False

    return True

Проверяет корректность разбиения вершин на два подмножества (set1 и set2). Для каждого элемента проверяется, есть ли связь между элементами внутри одного множества.
Также проверяется условие, что начальная вершина (start) и конечная вершина (end) находятся в разных множествах.

Алгоритм поиска всех возможных разрезов в графе

In [43]:
def find_all_cuts(graph, start, end, path=[]):
    list1=[]
    list2=[]
    for i in range(1,int(end)+1):
        list1.append(i)
    all_partitions = []
    for i in range(1, len(list1)):
        partitions = list(combinations(list1, i))
        all_partitions.extend(partitions)
    for partition in all_partitions:
        part1 = list(partition)
        part2 = [x for x in list1 if x not in part1]

        if(is_correct(part1,part2,graph,start,end)):
            list2.append([part1,part2])

    cuts=[]
    for i in list2:
        cut=[]
        for j in i[0]:
            for k in i[1]:
                if(str(k) in graph[str(j)] or str(j) in graph[str(k)]):
                    cut.append([str(j),str(k)])
        cuts.append(cut)
    return cuts

Находит все возможные разрезы графа, разделяющие его на два подмножества вершин таким образом, чтобы начальная и конечная вершины находились в разных подмножествах.
Для каждого корректного разбиения на два множества ищутся рёбра, пересекающие эти множества (которые и составляют разрез).

In [44]:
graph_matrix={}

In [45]:
graph = {
    '1': ['2','3'],
    '2': ['3','6'],
    '3': ['2', '4','5'],
    '4': ['5'],
    '5': ['6'],
    '6': []
}
weights=dict()
k=1;
for i in graph.items():
    for j in i[1]:
        if(weights.get(str(j)+' '+str(i[0]))==None):
            weights[str(i[0])+' '+str(j)]='x'+str(k)
            k+=1
print(weights)
start_node = '1'
end_node = '6'
paths=find_all_cuts(graph, start_node, end_node)
print(paths)
min_paths=[]
for i in paths:
    min_path=[]
    for j in i:
        if(weights.get(j[0]+' '+j[1])!=None):
            min_path.append(weights[j[0]+' '+j[1]])
        else:
            min_path.append(weights[j[1] + ' ' + j[0]])
    min_paths.append(min_path)
min_cuts = remove_duplicate_sublists(min_paths)
print(min_cuts)
S=1


for i in min_cuts:
    prod=1
    for j in i:
        prod*=1-sym.Symbol(j)
    prod=1-prod
    S*=prod

print(S)
S=sym.expand(S)

for k in range(10):
    for i in range(len(min_cuts)):
        for j in range(len(weights.items())):
            S = S.subs(sym.Symbol('x'+str(j+1))**(i+1), sym.Symbol('x'+str(j+1)))
print(S)
p=[0.845354,0.994669,0.996987,0.222299,0.98433,0.761534,0.775459,0.647558]
for i in range(1,len(weights.items())+1):
    S = S.subs(sym.Symbol('x' + str(i)), p[i-1])
print(S)

{'1 2': 'x1', '1 3': 'x2', '2 3': 'x3', '2 6': 'x4', '3 4': 'x5', '3 5': 'x6', '4 5': 'x7', '5 6': 'x8'}
[[['1', '2'], ['1', '3']], [['6', '2'], ['6', '5']], [['1', '3'], ['2', '3'], ['2', '6']], [['1', '2'], ['3', '2'], ['3', '4'], ['3', '5']], [['2', '1'], ['2', '3'], ['6', '5']], [['5', '3'], ['5', '4'], ['6', '2']], [['2', '6'], ['3', '4'], ['3', '5']], [['1', '2'], ['3', '2'], ['3', '5'], ['4', '5']], [['2', '1'], ['2', '3'], ['5', '3'], ['5', '4']], [['4', '3'], ['5', '3'], ['6', '2']], [['2', '6'], ['3', '5'], ['4', '5']], [['1', '2'], ['3', '2'], ['5', '6']], [['2', '1'], ['2', '3'], ['4', '3'], ['5', '3']], [['3', '1'], ['3', '2'], ['6', '2']], [['2', '6'], ['5', '6']], [['2', '1'], ['3', '1']]]
[['x1', 'x2'], ['x4', 'x8'], ['x2', 'x3', 'x4'], ['x1', 'x3', 'x5', 'x6'], ['x1', 'x3', 'x8'], ['x4', 'x6', 'x7'], ['x4', 'x5', 'x6'], ['x1', 'x3', 'x6', 'x7']]
(-(1 - x1)*(1 - x2) + 1)*(-(1 - x4)*(1 - x8) + 1)*(-(1 - x1)*(1 - x3)*(1 - x8) + 1)*(-(1 - x2)*(1 - x3)*(1 - x4) + 1)*(-(1 - 

Б) Пути

Рекурсивная функцию, которая находит все возможные пути в графе от стартовой вершины до конечной вершины

In [46]:
def find_all_paths(graph, start, end, path=[]):
    path = path + [start]
    if start == end:
        return [path]
    if start not in graph:
        return []
    paths = []
    for node in graph[start]:
        if node not in path:
            new_paths = find_all_paths(graph, node, end, path)
            for new_path in new_paths:
                paths.append(new_path)
    return paths

Эта рекурсивная функция находит все возможные пути из вершины start в вершину end в графе.
Она собирает все пути, проходя по графу в глубину (аналог поиска в глубину), избегая повторного посещения узлов в одном пути, чтобы избежать циклов.

In [47]:
graph_matrix={}

graph = {
    '1': ['2','3'],
    '2': ['3','6'],
    '3': ['2', '4','5'],
    '4': ['5'],
    '5': ['6'],
    '6': []
}
weights=dict()
k=1;
for i in graph.items():

    for j in i[1]:
        if(weights.get(str(j)+' '+str(i[0]))==None):
            weights[str(i[0])+' '+str(j)]='x'+str(k)
            k+=1
print(weights)
start_node = '1'
end_node = '6'
paths=find_all_paths(graph, start_node, end_node)
print(paths)
min_paths=[]
for i in paths:
    min_path=[]
    for j in range(len(i)-1):
        if(weights.get(i[j]+' '+i[j+1])!=None):
            min_path.append(weights[i[j]+' '+i[j+1]])
        else:
            min_path.append(weights[i[j+1] + ' ' + i[j]])
    min_paths.append(min_path)
print(min_paths)
S=1;
for i in min_paths:
    prod=1
    for j in i:
        prod*=sym.Symbol(j)
    prod=1-prod
    S*=prod
S=1-S
print(S)
S=sym.expand(S)
print(S)
for k in range(10):
    for i in range(len(min_paths)):
        for j in range(len(weights.items())):
            S = S.subs(sym.Symbol('x'+str(j+1))**(i+1), sym.Symbol('x'+str(j+1)))
print(S)
p=[0.845354,0.994669,0.996987,0.222299,0.98433,0.761534,0.775459,0.647558]

for i in range(1,len(weights.items())+1):
    S = S.subs(sym.Symbol('x' + str(i)), p[i-1])
print(S)

{'1 2': 'x1', '1 3': 'x2', '2 3': 'x3', '2 6': 'x4', '3 4': 'x5', '3 5': 'x6', '4 5': 'x7', '5 6': 'x8'}
[['1', '2', '3', '4', '5', '6'], ['1', '2', '3', '5', '6'], ['1', '2', '6'], ['1', '3', '2', '6'], ['1', '3', '4', '5', '6'], ['1', '3', '5', '6']]
[['x1', 'x3', 'x5', 'x7', 'x8'], ['x1', 'x3', 'x6', 'x8'], ['x1', 'x4'], ['x2', 'x3', 'x4'], ['x2', 'x5', 'x7', 'x8'], ['x2', 'x6', 'x8']]
-(-x1*x4 + 1)*(-x2*x3*x4 + 1)*(-x2*x6*x8 + 1)*(-x1*x3*x6*x8 + 1)*(-x2*x5*x7*x8 + 1)*(-x1*x3*x5*x7*x8 + 1) + 1
-x1**3*x2**3*x3**3*x4**2*x5**2*x6**2*x7**2*x8**4 + x1**3*x2**2*x3**3*x4**2*x5**2*x6*x7**2*x8**3 + x1**3*x2**2*x3**3*x4**2*x5*x6**2*x7*x8**3 + x1**3*x2**2*x3**2*x4*x5**2*x6**2*x7**2*x8**4 - x1**3*x2*x3**3*x4**2*x5*x6*x7*x8**2 - x1**3*x2*x3**2*x4*x5**2*x6*x7**2*x8**3 - x1**3*x2*x3**2*x4*x5*x6**2*x7*x8**3 + x1**3*x3**2*x4*x5*x6*x7*x8**2 + x1**2*x2**3*x3**3*x4*x5**2*x6**2*x7**2*x8**4 + x1**2*x2**3*x3**2*x4**2*x5**2*x6*x7**2*x8**3 + x1**2*x2**3*x3**2*x4**2*x5*x6**2*x7*x8**3 - x1**2*x2**2*x3**3*x4*x